### ISSUE TITLE
UI5按钮渲染原理

### ISSUE THREAD

**Issue by i042416** [i042416/KnowlegeRepository#2462]
To debug how is a button rendered in UI5 I created a project by using SAPUI5 application template via Web IDE. And add a button in View1.view.xml as below:

![clipboard1](https://user-images.githubusercontent.com/5669954/59420715-f77f4200-8dff-11e9-9979-09a7bd82899c.png)

Then I added a breakpoint in ButtonRender.js to check how a Button is rendered in UI5.
In below picture we can see the render sequence of the controls:
UIArea -> Shell -> UIComponent -> XMLView -> NavContainer -> PageRender -> ButtonRender
After we execute code of line 198, the button's text is not displayed on the page, so we know that we need to debug further.

![clipboard2](https://user-images.githubusercontent.com/5669954/59420717-f77f4200-8dff-11e9-91aa-5f8f2b74228e.png)

In writeEscape() function, we can see the button's text has been pushed in a buffer.

![clipboard3](https://user-images.githubusercontent.com/5669954/59420720-f817d880-8dff-11e9-9f88-10afb793b3b6.png)

And UI5 not calls DOM native method to draw controls one by one, UI5's strategy is rendering a batch of controls in one UIArea one time.

![clipboard4](https://user-images.githubusercontent.com/5669954/59420722-f817d880-8dff-11e9-9708-fa0f9365b4cc.png)

UI5 concatenates html elements which need to be drawn in a long String.

![clipboard5](https://user-images.githubusercontent.com/5669954/59420723-f8b06f00-8dff-11e9-9ea1-77922e965f29.png)

And then call append() function to draw these html elements.

![clipboard6](https://user-images.githubusercontent.com/5669954/59420724-f8b06f00-8dff-11e9-8f98-de2f650181f2.png)

jTarget is a JQuery object.

![clipboard7](https://user-images.githubusercontent.com/5669954/59420726-f8b06f00-8dff-11e9-84ca-109be97117cb.png)

In append() function call DOM native function appendChild() to draw html elements.

![clipboard8](https://user-images.githubusercontent.com/5669954/59420727-f9490580-8dff-11e9-9dbb-87a0887a6800.png)
![clipboard9](https://user-images.githubusercontent.com/5669954/59420728-f9490580-8dff-11e9-8ee1-f5f7a6b3a6ea.png)

As appendChild() is DOM native function we no need to jump into it, just step over line 10612, we can see that the button is displayed.

![clipboard10](https://user-images.githubusercontent.com/5669954/59420729-f9e19c00-8dff-11e9-936f-562b67a00608.png)

But the layout is strange, it is not display as we expected. we need to debug further. I continue to debug and step over many line's of code, the html elements' CSS style are still not change until I go to line 9184:

![clipboard11](https://user-images.githubusercontent.com/5669954/59420731-fa7a3280-8dff-11e9-9eb8-7aa618936d40.png)

And I enter the two parameters of jQuery.event.dispatch.apply() in console, we can see that the two parameters are like below:

![clipboard12](https://user-images.githubusercontent.com/5669954/59420732-fa7a3280-8dff-11e9-8203-cac48836cc6f.png)

If we set breakpoint in EventProvider.prototype.fireEvent() function, the code doesn't go there, that is to say this event is not a UI5 event.
In other words it will fire DOM native "load" event on DOM element "body". And after that we can see the all the DOM elements' including the button element's CSS style is correct as we expected. To here, I think I have made clear of how is a button rendered in UI5.

![clipboard13](https://user-images.githubusercontent.com/5669954/59420733-fb12c900-8dff-11e9-9b8b-aed7bc07b698.png)

要获取更多Jerry的原创文章，请关注公众号"汪子熙":
![](https://upload-images.jianshu.io/upload_images/2085791-c4f58b9813a954cb.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
